source/agora/common/Metadata.d

```diff
@@ -0,0 +1,137 @@
+/*******************************************************************************
+
+    Defines metadata which is stored to disk (e.g. peer info)
+
+    Copyright:
+        Copyright (c) 2019 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.Metadata;
+
+import agora.common.Data;
+import agora.common.Set;
+
+import vibe.core.log;
+import vibe.data.json;
+
+import std.array;
+import std.file;
+import std.format;
+import std.meta;
+import std.stdio;
+import std.typecons;
+
+/*******************************************************************************
+
+    Contains the serialized set of peers a node has connected to.
+    Note: Cannot be abstract, vibe.d requirement.
+
+*******************************************************************************/
+
+public class Metadata
+{
+    /// The set of peers we previously established connection to
+    public Set!Address peers;
+
+    /// Deserialize the Metadata from a Json string
+    public static void fromJsonString (Metadata metadata, string json)
+    {
+        metadata.peers = deserializeJson!(Set!Address)(json);
+    }
+
+    /// Dump the Metadata to a Json string
+    public string toJsonString ()
+    {
+        auto app = appender!string();
+        serializeToJson(app, this.peers);
+        return app.data[];
+    }
+
+    /// Load the metadata
+    public void load() { }
+
+    /// Dump the metadata
+    public void dump() { }
+}
+
+/// Metadata stored in memory (for unittests)
+public class MemMetadata : Metadata
```

Comment:
I think this should live in the `Base` module.
One thing I really try to avoid is mixing test code with "regular code".
There's two reason for that:
- Better separation of concern. The "regular" code should not know about the test code, but the other way around is fine.
- The regular code should not have any special "unittest" path. Any such path is a special behavior that might break one of the class' assumption.

Take the example of the `version (unittest)` constructor: having a different constructor might break some assumptions the class makes (now or later), by -e.g.- leaving some fields empty.